Add generated documentation

dalloriam opened the pull request:
This adds HTML API documentation generated from source with [Doxygen](https://github.com/doxygen/doxygen).

To best review it, check out this branch and open up `/docs/html/index.html` locally.

dalloriam committed:
Add generated documentation + stop generating latex

dalloriam committed:
Add initial github actions configuration

dalloriam committed:
Remove CMake install step

dalloriam committed:
Bump to ubuntu 20.04

dalloriam committed:
Minor documentation revisions

jpmarceau commented on README.md:
```diff
@@ -42,18 +42,23 @@ int main()
 ## Design Overview
 ### Background
-At its core, a cache has much in common with a hash map - both are datastructures used for associating a key with a value.
+At its core, a cache has much in common with a hash map; both are datastructures used for associating a key with a value.
```
```suggestion
At its core, a cache has much in common with a hash map; both are data structures used for associating a key with a value.
```

jpmarceau commented on README.md:
```diff
@@ -42,18 +42,23 @@ int main()
+A **caching scheme** is the set of algorithms used to select which items should be kept in cache over others.
+Since many different metrics can be prioritized when implementing a cache, there is no _one-size-fits-all_ caching scheme (e.g. a CPU cache usually only needs to prioritize the cache hit ratio, while a web server cache also needs to consider latency on cache miss).
+
+Because of this impossiblity to get a single well-performing and reusable caching scheme, many organizations and products fall back on using a simple [Least-Recently Used (LRU)](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_recently_used_(LRU)) cache everywhere. Although LRU is certainly _suitable_ for most use cases it is far from optimal, and using it without an afterthought mi ...
```
```suggestion
Because of this impossibility of relying on a single well-performing and reusable caching scheme, many organizations and products fall back on using a simple [Least-Recently Used (LRU)](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_recently_used_(LRU)) cache everywhere. Although LRU is certainly _suitable_ for most use cases, it is far from optimal and using it without an afterthought might lead to excessive memory usage or degraded performance.
```

jpmarceau commented on README.md:
```diff
@@ -42,18 +42,23 @@ int main()
+
+Because of this impossiblity to get a single well-performing and reusable caching scheme, many organizations and products fall back on using a simple [Least-Recently Used (LRU)](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_recently_used_(LRU)) cache everywhere. Although LRU is certainly _suitable_ for most use cases it is far from optimal, and using it without an afterthought mi ...
+
+Cachemere is designed with the explicit goal of providing a single reusable cache that can be customized to use different schemes, allowing applications to get the maintability benefits of a single cache implementation, while at the same time getting all the performance benefits of a purpose-built cache.
```
```suggestion
Cachemere is designed with the explicit goal of providing a single reusable cache that can be customized using different schemes, allowing applications to get the maintainability benefits of a single cache implementation, while at the same time getting all the performance benefits of a purpose-built cache.
```

jpmarceau commented on README.md:
```diff
@@ -42,18 +42,23 @@ int main()
+Cachemere tackles this goal in a modular fashion with the concept of _policies_. A `cachemere::Cache` is parameterized
 by an **Insertion Policy** and an **Eviction Policy**.
 
 Broadly speaking, the job of an **Insertion Policy** is to determine whether an item should be inserted or kept in cache, while the job of an **Eviction
```
In the next line:
`to determine which item(s) should be evicted from the cache to make room for an new item.`
->
`to determine which item(s) should be evicted from the cache.`

jpmarceau commented on README.md:
```diff
@@ -63,9 +68,9 @@ decision on whether an item should be added or excluded.
-For instance, if instead of using more common policies like Least-Recently Used (LRU)
-or Least-Frequently Used (LFU), you wanted to implement a custom cache that takes the frequency of access **as well as** the size of the item into
-consideration when establishing which item to evict, you could simply implement a `SizeBasedEvictionPolicy` and use it with the existing `cachemere::Cache`.
+For instance, if instead of using more common policies like Least-Recently Used (LRU),
```
```suggestion
For instance, instead of using more common policies like Least-Recently Used (LRU),
```

jpmarceau commented on README.md:
```diff
@@ -63,9 +68,9 @@ decision on whether an item should be added or excluded.
-or Least-Frequently Used (LFU), you wanted to implement a custom cache that takes the frequency of access **as well as** the size of the item into
-consideration when establishing which item to evict, you could simply implement a `SizeBasedEvictionPolicy` and use it with the existing `cachemere::Cache`.
+For instance, if instead of using more common policies like Least-Recently Used (LRU),
+a product required the use of a custom cache that takes the frequency of access as well as the size of the item into
```
```suggestion
a product might need to use a custom cache that takes the frequency of access as well as the size of the item into
```

jpmarceau commented on README.md:
```diff
@@ -63,9 +68,9 @@ decision on whether an item should be added or excluded.
-consideration when establishing which item to evict, you could simply implement a `SizeBasedEvictionPolicy` and use it with the existing `cachemere::Cache`.
+For instance, if instead of using more common policies like Least-Recently Used (LRU),
+a product required the use of a custom cache that takes the frequency of access as well as the size of the item into
+consideration when establishing which item to evict. To solve this, one could implement a `SizeBasedEvictionPolicy` and use it with the existing `cachemere::Cache` without issues.
```
```suggestion
consideration when establishing which item to evict. To do this, one could implement a `SizeBasedEvictionPolicy` and use it with the existing `cachemere::Cache`.
```

jpmarceau replied:
(fyi, removing the `without issues` part is debatable, but it's generally a good practice never to imply that something is `simple` when writing documentation)

jpmarceau commented on README.md:
```diff
@@ -63,9 +68,9 @@ decision on whether an item should be added or excluded.
+a product required the use of a custom cache that takes the frequency of access as well as the size of the item into
+consideration when establishing which item to evict. To solve this, one could implement a `SizeBasedEvictionPolicy` and use it with the existing `cachemere::Cache` without issues.
 
 Similarly, if a use case required the cache to be constrained by its number of items instead of by the amount of memory it uses (which is the current
```
Here's a suggestion for this paragraph:
`Similarly, imagine a cache that limits the amount of memory it uses. Now, imagine a use case that additionally requires constraining the number of items in the cache. To handle this scenario, you could implement a new `Cache` object parametrized by these two policies and reuse Cachemere's LRU implementation of this `Cache` object.`

dalloriam replied:
How about 

`Similarly, imagine a cache that limits the amount of memory it uses. Now, imagine a use case that additionally requires constraining the number of items in the cache. To handle this scenario, you could implement a new Cache object parameterized by these two policies and reuse Cachemere's LRU policy with this Cache object.` 

I feel like this does a better job of conveying which component of the library could be reused with the user implementation

jpmarceau replied:
Sounds good!

jpmarceau commented on implementingCustomPolicies.dox:
```diff
@@ -1,7 +1,7 @@
 ============================
 
-This page is about implementing custom policies.
+This page is about guiding the implementation of custom policies.
```
```suggestion
This page is meant to guide you through the implementation of custom policies.
```

jpmarceau commented on implementingCustomPolicies.dox:
```diff
@@ -1,7 +1,7 @@
+This page is about guiding the implementation of custom policies.
 ## Shared Characteristics
 ### Construction Requirements
 Both types of policies _must_ be default-constructible, because the cache is responsible for instanciating them.
```
It's been a while since I've worked with C or C++, so maybe I'm just rusty, but it could be worth linking to some documentation about [default-constructibility](http://www.cplusplus.com/reference/type_traits/is_default_constructible/).

jpmarceau replied:
Below, on line 9, I'm not sure what you mean by 
`These handlers will only be called if implemented.`
Do you mean that the user has to implement the handler altogether? When you write `will only be called`, do you mean that some part of the package will call them if they exist?

jpmarceau replied:
Below, at like 53
`if the policy believes that the candidate should`
->
`if the policy states that the candidate should`
(believes sounds a bit too human I think)

dalloriam replied:
> Below, on line 9, I'm not sure what you mean by
> `These handlers will only be called if implemented.`
> Do you mean that the user has to implement the handler altogether? When you write `will only be called`, do you mean that some part of the package will call them if they exist?

Basically, the `Cache` class does some template magic to figure out which event handlers are defined on the policy, and only generates the code to call the ones that are defined.

How about 
>  These handlers do not all need to be implemented; a policy only needs to define the handlers that are relevant to its implementation. The cache will detect available handlers on its policies at compile-time, and will only call those that are defined.

jpmarceau replied:
Yep that's clear thank you 👍

jpmarceau commented on docs/html/classcachemere_1_1Cache.html:
```diff
@@ -0,0 +1,485 @@
+</table>
+</div><div class="memdoc">
+
+<p>Check if a given key is stored in the cache. </p>
```
```suggestion
<p>Check whether a given key is stored in the cache. </p>
```

jpmarceau replied:
(I realize that this is generated from the `cache.h` file, but it is not included in this PR, so I think this is the best place for me to comment 😅

jpmarceau commented on docs/html/classcachemere_1_1Cache.html:
```diff
@@ -0,0 +1,485 @@
+<p>Check if a given key is stored in the cache. </p>
+<dl class="params"><dt>Parameters</dt><dd>
+  <table class="params">
+    <tr><td class="paramname">key</td><td>The key whose presence needs to be tested. </td></tr>
```
```suggestion
    <tr><td class="paramname">key</td><td>The key whose presence to test. </td></tr>
```

jpmarceau commented on docs/html/classcachemere_1_1Cache.html:
```diff
@@ -0,0 +1,485 @@
+</div><div class="memdoc">
+
+<p>Get the number of items currently stored in the cache. </p>
+<dl class="section warning"><dt>Warning</dt><dd>This method acquires an internal mutex to grab the number of items. Calling this excessively while the cache is under contention will be detrimental to performance. </dd></dl>
```
What is an internal mutex?
If it's not obvious to readers, I would link to documentation on that subject :)

dalloriam replied:
The `internal` qualifiier is probably superfluous. I think mutex is pretty clear, but I'll change it to `mutual exclusion lock` to be  more explicit :smile:

jpmarceau commented on docs/html/classcachemere_1_1Cache.html:
```diff
@@ -0,0 +1,485 @@
+<p>Get the amount of memory currently being used by cache items. </p>
+<p>This method returns the size of the data stored by the cache itself:</p><ul>
+<li>The size of the key and the value, as defined by the <code>MeasureKey</code> and <code>MeasureValue</code> functors</li>
+<li>A copy of the key <dl class="section return"><dt>Returns</dt><dd>The current memory usage, in bytes. </dd></dl>
```
hmmm... there seems to be something weird here. You describe what gets returned but they you have a list where you mention `A copy of the key` and then again you have a `Returns` section 🤔 
<img width="713" alt="Screen Shot 2021-03-16 at 9 33 23 AM" src="https://user-images.githubusercontent.com/39384459/111317620-8ea2aa80-8621-11eb-8efc-5a2b0fa6cfcd.png">

dalloriam replied:
I agree this needs to be clearer. What I meant is that the _total_ memory usage is the sum of the size of the content inserted by the user (key + value) **plus** the size of a copy of the key that is kept by the cache for internal purposes.

I'll improve that part!

jpmarceau commented on docs/html/classcachemere_1_1policy_1_1EvictionLRU.html:
```diff
@@ -0,0 +1,297 @@
+      </table>
+</div><div class="memdoc">
+
+<p>Get an end iterator. </p>
```
What's an "end iterator" exactly? It always returns the last element that would become the victim?

dalloriam replied:
An [end iterator](https://www.cplusplus.com/reference/iterator/end/) is an iterator pointing _one past the end_ of the sequence. I'd consider this to be a well-known concept for most C++ developers (as most containers in the standard library conform to this API), although a link couldn't hurt :smile: 

As for 
> It always returns the last element that would become the victim?

I think something is not quite clear in the docs.
The LRU and Segmented LRU policies internally store keys from hottest to coldest. This means that the last item in the list is the one that will be evicted.

However the policy interface states that `victim_begin` must yield an iterator to _the first item that should be evicted_, so effectively, in the case of LRU-based policies, the iterator returned will walk the internal structure in reverse order.
I think should be clarified by adding a comment at the class level for both LRU policies that indicates that the policy internally stores keys from hottest to coldest, as well as a comment on `victim_begin` stating that the victim iterator will walk the internal structure backwards - what do you think?

jpmarceau replied:
ah that makes sense
And yes, I think quick notes would be useful for readers to clarify the directions of the data structures, because it can be counterintuitive when you first look at it :)

jpmarceau commented on docs/html/classcachemere_1_1policy_1_1EvictionLRU.html:
```diff
@@ -0,0 +1,297 @@
+<p>Least Recently Used (LRU) eviction policy. </p>
+<p>Implemented using an internal linked list. Only stores references to keys kept alive by the cache. </p><dl class="tparams"><dt>Template Parameters</dt><dd>
+  <table class="tparams">
+    <tr><td class="paramname">Key</td><td>The key used to identify items in the cache. </td></tr>
```
Why do we go from singular `key` to plural `items`? There must be something I don't get here 🤔

dalloriam replied:
Right. These are template type parameters, so the correct comment would be `The type of the keys used to identify items in the cache.`

jpmarceau commented on docs/html/classcachemere_1_1policy_1_1EvictionLRU.html:
```diff
@@ -0,0 +1,297 @@
+<p>Implemented using an internal linked list. Only stores references to keys kept alive by the cache. </p><dl class="tparams"><dt>Template Parameters</dt><dd>
+  <table class="tparams">
+    <tr><td class="paramname">Key</td><td>The key used to identify items in the cache. </td></tr>
+    <tr><td class="paramname">Value</td><td>The values stored in the cache. </td></tr>
```
Here as well, I'm not sure why we talk about `values` plural?

dalloriam replied:
Similarly here, it'd be `The type of the values stored in the cache`

jpmarceau commented on docs/html/classcachemere_1_1policy_1_1EvictionLRU.html:
```diff
@@ -0,0 +1,297 @@
+</div><div class="memdoc">
+
+<p><a class="el" href="classcachemere_1_1Cache.html" title="Thread-safe memory-restricted cache.">Cache</a> hit event handler. </p>
+<p>Moves the provided item at the front of the list. </p><dl class="params"><dt>Parameters</dt><dd>
```
What do you mean by `front of the list`? Presumably this would be the last one to get evicted?

dalloriam replied:
I suppose this comment and the next were answered by my previous comment about LRU ordering? :smile:

jpmarceau commented on docs/html/classcachemere_1_1policy_1_1EvictionLRU.html:
```diff
@@ -0,0 +1,297 @@
+</div><div class="memdoc">
+
+<p>Insertion event handler. </p>
+<p>Inserts the provided item at the front of the list. </p><dl class="params"><dt>Parameters</dt><dd>
```
I'm a bit confused here. The event handler puts the provided item at the front of the list?
I'm similarly confused by the `Update event handler` below. 
Is there something I'm not seeing here?

jpmarceau commented on docs/html/classcachemere_1_1policy_1_1detail_1_1BloomFilter.html:
```diff
@@ -0,0 +1,270 @@
+<p>Probabilistic datastructure for representing sets in a space-efficient format. </p>
+<p>A bloom filter is a constant-sized datastructure, which means that insertions will never make the filter allocate more memory. However, too many inserts will severly impact the accuracy of filter membership tests. </p><dl class="tparams"><dt>Template Parameters</dt><dd>
+  <table class="tparams">
+    <tr><td class="paramname">Key</td><td>The type of the items that will be inserted into the set. </td></tr>
```
Is it normal that the paramname `Key` doesn't seem to match very well with its description?
I would've expected something like `KeyType` 🤔

dalloriam replied:
Great point. 
Actually, the name `Key` isn't super appropriate for this data structure - it makes a bit of sense because TinyLFU uses it to store keys, but it technically could be used to store any hashable type. I'll change the naming all around to `item`, and change the template parameter to `Item` (`Item` instead of `ItemType` because since it's a `typename` template parameter, we already know it's a type)

jpmarceau commented on docs/html/classcachemere_1_1policy_1_1detail_1_1BloomFilter.html:
```diff
@@ -0,0 +1,270 @@
+</div><div class="memdoc">
+
+<p>Constructor. </p>
+<p>To use this datastructure at its full potential, it's very important to have a good estimate for the cardinality of the set to be inserted.</p>
```
```suggestion
<p>To use this data structure to its full potential, it's very important to have a good estimate for the cardinality of the set to be inserted.</p>
```

jpmarceau commented on docs/html/classcachemere_1_1policy_1_1detail_1_1BloomFilter.html:
```diff
@@ -0,0 +1,270 @@
+      </table>
+</div><div class="memdoc">
+
+<p>Add an item to the filter. </p>
```
Should it rather be `Add a key to the filter`?

jpmarceau commented on docs/html/classcachemere_1_1Cache.html:
```diff
@@ -0,0 +1,485 @@
+<p>This class keeps the inserted items alive, and it handles the bulk of the Insert/Evict logic while respecting the size constraints.</p>
+<p>Some logic is delegated to the insertion and eviction policies. For details, see the <a class="el" href="index.html">Main Page</a>. </p><dl class="tparams"><dt>Template Parameters</dt><dd>
+  <table class="tparams">
+    <tr><td class="paramname">Key</td><td>The type of the key used for retrieving items. </td></tr>
```
Here as well I would expect `KeyType` rather than `Key`

dalloriam replied:
Similar to my comment in the bloom filter, I feel like `Key` (and `Value`) is appropriate here, given that these are `typename` template parameters, but if you'd really prefer I can switch it over
![image](https://user-images.githubusercontent.com/1795950/111375044-31841480-8674-11eb-9a8c-eb03e532f2ba.png)

jpmarceau commented on docs/html/classcachemere_1_1Cache.html:
```diff
@@ -0,0 +1,485 @@
+<p>Some logic is delegated to the insertion and eviction policies. For details, see the <a class="el" href="index.html">Main Page</a>. </p><dl class="tparams"><dt>Template Parameters</dt><dd>
+  <table class="tparams">
+    <tr><td class="paramname">Key</td><td>The type of the key used for retrieving items. </td></tr>
+    <tr><td class="paramname">Value</td><td>The type of the items stored in the cache. </td></tr>
```
And here `ValueType`

jpmarceau commented on docs/html/classcachemere_1_1Cache.html:
```diff
@@ -0,0 +1,485 @@
+</table>
+</div><div class="memdoc">
+
+<p>Compute and return the running byte hit ratio of the cache, in bytes. </p>
```
A ratio in bytes? Don't units typically cancel out?
5B/10B = 5/10?

dalloriam replied:
_Technically_ the proper name is [byte hit rate](https://www.igi-global.com/dictionary/web-caching/36973), i should change that for googleability.

Nomenclature aside, as stated in the `details` section further down:
> The byte hit ratio represents the average amount of data saved by cache accesses.

The metric _is_ a ratio, but not of bytes over bytes. It's bytes over access count:

(25b + 25b + 0b) / 3 = 16.67b

jpmarceau commented on docs/html/classcachemere_1_1Cache.html:
```diff
@@ -0,0 +1,485 @@
+      </table>
+</div><div class="memdoc">
+
+<p>Find a given key in cache, optionally returning the associated value. </p>
```
```suggestion
<p>Find a given key in cache, returning the associated value when it exists. </p>
```

jpmarceau reviewed (commented):
Thanks for thinking about us and asking for a review 😄 

At times I left comments in the generated html files because the originals were not available in this PR. I hope it's understandable but let me know if you have questions or if you think of a better way to do this. In a number of cases I left questions so I will likely make other suggestions once I have answers.

The content is very good, so my comments are always targeted at precise formulations. Great idea to put up an example right away too, it's the main thing for users to get started! Nice job 👌

dalloriam committed:
Address documentation comments

dalloriam commented:
@jpmarceau Just pushed all requested edits. Thanks a lot for the thorough review! :)

jpmarceau reviewed (approved):
Looks great!

dalloriam committed:
Merge branch 'feature/generated-docs' into feature/ci

dalloriam committed:
Add GDSF eviction policy + new preset for custom cost caching

dalloriam committed:
Self-review

dalloriam committed:
Add generated docs for GDSF

dalloriam committed:
Add basic performance + accuracy benchmarks

dalloriam committed:
Add TBB to github actions setup

dalloriam committed:
Add google benchmark setup to CI

dalloriam committed:
Add removal-related utility functions

dalloriam committed:
Add swap and collect_into

dalloriam committed:
Add a way to disable cache locking at compile-time

dalloriam committed:
Run tests on both GCC and clang

dalloriam committed:
Using an actual C++ compiler would help

dalloriam committed:
Fix review comments & fix nasty iterator bug in SLRU

dalloriam committed:
Add victim_iterator UTs to the LRU and GDSF policies

dalloriam committed:
Add DoNotOptimize to find() call

dalloriam committed:
Add first vcpkg test

dalloriam committed:
Use bash shell for building on windows

dalloriam committed:
Add cmake policy for vcpkg gtest

dalloriam committed:
Add policy before project directive

dalloriam committed:
Play nice with MSVC

dalloriam committed:
Fix executable extension

dalloriam committed:
Use bash for running tests also

dalloriam committed:
Merge pull request #6 from coveooss/feature/IDX-3749-prerequisite-utilities

[IDX-3749] - Add cache replacement prerequisites

dalloriam committed:
Merge pull request #5 from coveooss/feature/benchmarks

[IDX-3730] - Performance & Accuracy Benchmarks

dalloriam committed:
Merge branch 'feature/ci' into feature/custom-cost-caching

dalloriam committed:
Fix linking against TBB

dalloriam committed:
Remove template instanciation in decltype to dodge possible msvc bug

dalloriam committed:
Merge pull request #4 from coveooss/feature/custom-cost-caching

[IDX-3725] - Add GDSF eviction policy + new preset for custom cost caching

dalloriam committed:
Hide UT and benchmark compilation behind feature flags

dalloriam committed:
Fix find_package calls with vcpkg, add -Wall, -Wextra, and /W4

dalloriam committed:
Hopefully fix find_boost

dalloriam committed:
Remove config mode for boost

klalumiere committed:
Removing CMake files

This will allow me to add them back, and hence to review them more
easily using GitHub tooling.

klalumiere committed:
Adding CMake files back

This allows me to review them more easily using GitHub tooling.

dalloriam committed:
Address review comments

dalloriam committed:
Add missing indent

dalloriam committed:
Merge pull request #7 from coveooss/code_review/cmake-adding-back

Code Review CMake files

dalloriam committed:
Merge pull request #3 from coveooss/feature/ci

Add github action to run tests on push
